### PR TITLE
Add more queue and task processing metrics

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Add task load latency (`task_latency_load`), which measures the task load time (from task generation to loading into memory). It also contains `queue_reader_id` tag for multi-cursor implementation, so that we can know which reader loads that task and only alert on metrics emit by reader id 0.
- Add task schedule latency (`task_latency_schedule`), which is the duration between task submission to task execution (basically, one time in memory queue latency)
- Add queue schedule latency (`queue_latency_schedule`), which measures the speed of task schedule poll tasks from task channel (throughput). The value is how long it takes task scheduler to schedule/poll 100 tasks from one task channel. The value is 100 task as the time may be too short (< 1ms) for 1 task. Operation should alert on this metric instead of task schedule latency as that one may increase due to user backlog.
- Add metrics for reader count, slice count, queue action.

**Why?**

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: service/history/queues/scheduler_monitor.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+func (m *schedulerMonitor) Stop() {
+	if !atomic.CompareAndSwapInt32(&m.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(m.shutdownCh)
+}
+
+func (m *schedulerMonitor) RecordStart(executable Executable) {
+	startTime := m.timeSource.Now()
```
Comment (yiminc) [inline service/history/queues/scheduler_monitor.go:122]: should we have startTime after obtain lock?

File: service/history/queues/scheduler_monitor.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	for {
+		select {
+		case <-m.shutdownCh:
+			return
+		case <-emissionTicker.C:
+			m.Lock()
+
+			now := m.timeSource.Now()
+			for taskChanKey, stats := range m.scheduleStats {
+				if now.Sub(stats.lastEmissionTime) > m.options.aggregationDuration {
```
Comment (yiminc) [inline service/history/queues/scheduler_monitor.go:156]: Some similarity with https://github.com/temporalio/temporal/pull/2970.
Maybe we could abstract some generic logic here as well (not in this PR).

File: service/history/queues/scheduler_monitor.go (hunk 1 above)
Comment (yycptt) [reply service/history/queues/scheduler_monitor.go:122]: By the time this method is invoked, the task has already been polled from the task channel and started processing, so I think the start time should be recorded before obtaining the lock (lock latency is not part of queue schedule latency).

File: service/history/queues/metrics.go [hunk 3]
```diff
@@ -57,6 +60,11 @@ const (
 	AckLevelUpdateCounter       = "ack_level_update"
 	AckLevelUpdateFailedCounter = "ack_level_update_failed"
 	PendingTasksCounter         = "pending_tasks"
+
+	QueueScheduleLatency      = "queue_latency_schedule" // latency for scheduling 100 tasks in one task channel
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/history/queues/metrics.go:64]: On [L202](https://github.com/temporalio/temporal/pull/3274/files#diff-295b97d8b186b39a45ef06657e00dd2b0e2bbfb99fcafcf1679f122666fb69bfR202) of `schduler_monitor.go`, when emitting the metrics, the value will be scaled to be 100 tasks. So if there's only 20 tasks in the last 30s, the emitted value will be 5 * total latency for those 20 tasks.

File: service/history/queues/executable.go [hunk 4]
```diff
@@ -303,14 +317,20 @@ func (e *executableImpl) Ack() {
 	e.state = ctasks.TaskStateAcked
 
 	if e.shouldProcess {
+		e.metricsProvider.Timer(TaskLoadLatency).Record(
+			e.loadTime.Sub(e.GetVisibilityTime()),
+			metrics.QueueReaderIDTag(e.readerID),
```
Comment (yiminc) [inline service/history/queues/executable.go:322]: It looks like the readerID is only needed for metrics tag. Would it be possible to tag the metricsHandler with the readerID before creating the executable so executable don't need to know about the readerID?

**Review by yiminc — APPROVED** [review]

File: service/history/queues/executable.go (hunk 4 above)
Comment (yycptt) [reply service/history/queues/executable.go:322]: I am thinking that readerID is only relevant for metrics that including db latency (`task_latency_queue{,_nouserlatency}` and `task_latency_load`). It's weird that metrics like `task_latency_processing` also contains the readerID tag, as it has nothing to do with task loading.  It also adds unnecessary cardinality for those metrics.

**Pull request merged**